controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CreateReaderGroupTask.java

```diff
@@ -0,0 +1,92 @@
+import io.pravega.controller.util.RetryHelper;
+import io.pravega.shared.NameUtils;
+import io.pravega.shared.controller.event.CreateReaderGroupEvent;
+import lombok.extern.slf4j.Slf4j;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.Iterator;
+
+/**
+ * Request handler for executing a create operation for a KeyValueTable.
+ */
+@Slf4j
+public class CreateReaderGroupTask implements ReaderGroupTask<CreateReaderGroupEvent> {
+
+    private final StreamMetadataStore streamMetadataStore;
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final ScheduledExecutorService executor;
+
+    public CreateReaderGroupTask(final StreamMetadataTasks streamMetaTasks,
+                                 final StreamMetadataStore streamMetaStore,
+                                 final ScheduledExecutorService executor) {
+        Preconditions.checkNotNull(streamMetaStore);
+        Preconditions.checkNotNull(streamMetaTasks);
+        Preconditions.checkNotNull(executor);
+        this.streamMetadataStore = streamMetaStore;
+        this.streamMetadataTasks = streamMetaTasks;
+        this.executor = executor;
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(final CreateReaderGroupEvent request) {
+        String scope = request.getScopeName();
+        String readerGroup = request.getRgName();
+        long requestId = request.getRequestId();
+
+        final RGOperationContext context = streamMetadataStore.createRGContext(scope, readerGroup);
+        log.info("Processing create request.");
+        return RetryHelper.withRetriesAsync(() -> streamMetadataStore.getVersionedReaderGroupState(scope, readerGroup,
+                true, context, executor)
+                .thenCompose(state -> {
+                    if (state.getObject().equals(ReaderGroupState.CREATING)) {
+                        String scopedRGName = NameUtils.getScopedReaderGroupName(scope, readerGroup);
+                        return streamMetadataStore.getReaderGroupConfigRecord(scope, readerGroup, context, executor)
+                               .thenCompose(configRecord -> {
+                               log.info("Adding metadata to Streams");
+                               if (!ReaderGroupConfig.StreamDataRetention.values()[configRecord.getObject().getRetentionTypeOrdinal()]
+                                   .equals(ReaderGroupConfig.StreamDataRetention.NONE)) {
+                                   // update Stream metadata tables, if RG is a Subscriber
+                                   Iterator<String> streamIter = configRecord.getObject().getStartingStreamCuts().keySet().iterator();
+                                   return Futures.loop(() -> streamIter.hasNext(), () -> {
+                                          Stream stream = Stream.of(streamIter.next());
+                                          return streamMetadataStore.addSubscriber(stream.getScope(),
```

Comment:
shouldnt add subscriber include generation number?

Reply:
There should not be a need to maintain generation number in Stream Metadata, if `updateSubscriberStreamCut()` takes a Reader Group `generation` as a parameter and checks the generation against the RGMetadata table before attempting to update the stream-cut in Stream metadata.

Reply:
Will add this as discussed.